primitives/src/erasure_code/reed_solomon_erasure.rs

```diff
@@ -0,0 +1,218 @@
+/// polynomial whose coefficients are the message entries. Decoding is a naive
+/// Lagrange interpolation.
+pub struct ReedSolomonErasureCode<F: PrimeField> {
+    reconstruction_size: usize,
+    num_shards: usize,
+    phantom_f: PhantomData<F>,
+}
+
+/// Shards for Reed Solomon erasure code
+#[derive(Clone, Eq, Debug, CanonicalSerialize, CanonicalDeserialize, Derivative)]
+#[derivative(PartialEq, Hash(bound = "F: PrimeField"))]
+pub struct ReedSolomonErasureCodeShard<F: PrimeField> {
+    /// Index of shard shard
+    pub index: usize,
+    /// Value of this shard
+    pub values: Vec<F>,
+}
+
+impl<F: PrimeField> ReedSolomonErasureCodeShard<F> {
+    /// Create a new shard
+    pub fn new(index: usize, values: Vec<F>) -> Self {
+        Self { index, values }
+    }
+}
+
+impl<F: PrimeField> ErasureCode for ReedSolomonErasureCode<F> {
+    type Field = F;
+    type Shard = ReedSolomonErasureCodeShard<F>;
+
+    fn new(reconstruction_size: usize, num_shards: usize) -> Result<Self, PrimitivesError> {
+        if reconstruction_size > num_shards {
+            Err(PrimitivesError::ParameterError(
+                "Number of shards must be at least the message length.".to_string(),
+            ))
+        } else {
+            Ok(ReedSolomonErasureCode {
+                reconstruction_size,
+                num_shards,
+                phantom_f: PhantomData,
+            })
+        }
+    }
+
+    /// The encoding will split the data into chunks of length
+    /// `reconstruction_size`. And represent each chunk as a polynomial. The
+    /// codeword composes of evaluations of those polynomials on
+    /// (1..num_shards).
+    fn encode(&self, data: &[Self::Field]) -> Result<Vec<Self::Shard>, PrimitivesError> {
+        let result = (1..self.num_shards + 1)
+            .map(|i| {
+                let mut vals = vec![];
+                let mut val = F::zero();
+                let mut x = F::one();
+                let mut cnt: usize = 0;
+                data.iter().for_each(|coef| {
+                    val += x * coef.borrow();
+                    x *= F::from(i as u64);
+                    cnt += 1;
+                    if cnt % self.reconstruction_size == 0 {
+                        vals.push(val);
+                        val = F::zero();
+                        x = F::one();
+                    }
+                });
+                if cnt % self.reconstruction_size != 0 {
+                    vals.push(val);
+                }
+                ReedSolomonErasureCodeShard::new(i, vals)
+            })
+            .collect::<Vec<_>>();
+        Ok(result)
+    }
+
+    /// Lagrange interpolation
+    /// Given a list of points (x_1, y_1) ... (x_n, y_n)
+    ///  1. Define l(x) = \prod (x - x_i)
+    ///  2. Calculate the barycentric weight w_i = \prod_{j \neq i} 1 / (x_i -
+    /// x_j)  
+    ///  3. Calculate l_i(x) = w_i * l(x) / (x - x_i)
+    ///  4. Return f(x) = \sum_i y_i * l_i(x)
```

Comment:
I'm confused. `encode` takes an arbitrary size `data` and partitions it into multiple polynomials of degree `reconstruction_size`. Yet this comment suggests that `decode` returns only a _single_ polynomial `f` of degree `reconstruction_size`. What happened to all the other polynomials?